developer_manual/app/extstorage.rst

```diff
@@ -1,67 +1,86 @@
 
-The :file:`/appinfo/info.xml` must be adjusted to specify **filesystem** as type:
+.. NOTE:: 
+   To save time, however, you can learn from an existing example, by reading
+   through the source code of the `FTP external storage app
+   <https://github.com/owncloud/files_external_ftp>`_.
+
+Configure the filesystem type
+=============================
+
+First, the :file:`/appinfo/info.xml` must be adjusted to specify the ``type`` as:
+``filesystem``. For example:
 
 .. code-block:: xml
 
 .. include:: examples/storage-backend/appinfo/info.xml
 
-Implementing the storage
-========================
+Implement the storage class(es)
+===============================
 
-Usually one should implement the interface **\\OCP\\Files\\Storage\\IStorage**
-but the easiest way is to directly extend
-**\\OCP\\Files\\Storage\\StorageAdapter** which already provides the
-implementation for many commonly required methods.
+Next, you need to create a storage class. Usually, you should implement the
+interface ``\\OCP\\Files\\Storage\\IStorage``. But, the easiest way is to
+directly extend ``\\OCP\\Files\\Storage\\StorageAdapter``, as it already
+provides an implementation for many of the commonly required methods.
 
-For this storage, we'll use a fake library which we'll call **FakeStorageLib**.
-We create a class that implements all filesystem operations required by
-ownCloud.
+Here’s an example of how you would create one that implements all the
+filesystem operations required by ownCloud, using a fictitious library called
+``FakeStorageLib``.
 
 .. code-block:: php
 
 .. include:: examples/storage-backend/OCA/MyStorageApp/Storage/MyStorage.php
 
-For this example we simply mapped the storage methods to the one from the
-library. In many cases, the underlying library might not support some
+For this example we mapped the available storage methods to the ones from the
+library. Note that, in many cases, the underlying library might not support some
 operations and might need extra code to work this around.
 
-Stat cache
-----------
+.. NOTE::
+   When extending StorageAdapter, it is best to also implement some methods
+   like ``rename``, and ``isCreatable`` etc, for performance reasons. If you
+   don’t, your storage backend will still work. But it will likely be slower,
+   because the default method would use a stream copy + delete for rename.
 
-Within a single PHP request, ownCloud might call the same storage methods
-repeatedly due to different checks. If the underlying library does not support
-stat/metadata caching, you might need to implement a stat cache yourself. For
-this, the metadata of any folder entries should be cached in a local array and
-returned by the storage methods.
+Stat/metadata cache
+-------------------
+
+To create a mature implementation, we need to consider stat and metadata
+caching. Within a single PHP request, ownCloud might call the same storage
+methods repeatedly, due to different checks which it need to carry out. As
```

Comment:
Thanks @phil-davis. Somehow I need to integrate [Grammarly](http://grammarly.com) with VIM